featurizers/fingerprints.py

```diff
@@ -1,68 +1,234 @@
-import multiprocessing as mp
 from typing import Union
 
 import numpy as np
 import pandas as pd
 import rdkit.Chem as Chem
-from rdkit.Chem.rdMolDescriptors import GetMorganFingerprintAsBitVect
 
 from base import FingerprintTransformer
 
+"""
+If during multiprocessing occurs MaybeEncodingError, first check if there isn't thrown any exception inside
+worker function! (That error isn't very informative and this tip might save you a lot of time)
+"""
 
-class MorganFingerprintAsBitVect(FingerprintTransformer):
+"""
+fp_descriptors needs to be here or inside _transform() of a specific class (cannot be defined inside the __init__() of
+that class), otherwise pickle gets angry:
+        TypeError: cannot pickle 'Boost.Python.function' object
+"""
+
+
+class MorganFingerprint(FingerprintTransformer):
     def __init__(
         self,
         radius: int = 2,
         n_bits: int = 2048,
         use_chirality: bool = False,
         use_bond_types: bool = True,
         use_features: bool = False,
+        result_type: str = "default",
         n_jobs: int = 1,
     ):
+        assert result_type in ["default", "as_bit_vect", "hashed"]
+
+        super().__init__(n_jobs)
         self.radius = radius
         self.n_bits = n_bits
         self.use_chirality = use_chirality
         self.use_bond_types = use_bond_types
         self.use_features = use_features
+        self.result_type = result_type
+
+    def _calculate_fingerprint(
+        self, X: Union[pd.DataFrame, np.ndarray]
+    ) -> np.ndarray:
+        fp_args = {
+            "radius": self.radius,
+            "useChirality": self.use_chirality,
+            "useBondTypes": self.use_bond_types,
+            "useFeatures": self.use_features,
+        }
 
-        if n_jobs == -1:
-            self.n_jobs = mp.cpu_count()
+        if self.result_type == "default":
+            from rdkit.Chem.rdMolDescriptors import GetMorganFingerprint
```

Comment:
W sumie chyba można by to przerobić, żeby korzystać z tego fajnego interfejsu "Generate": https://www.rdkit.org/docs/GettingStartedInPython.html#fingerprinting-and-molecular-similarity. Zróbcie na to osobny ticket, bo to jest improvement w kolejnym PRze

Reply:
Zapamiętamy